### PR TITLE
RPC: improve supporting tx execution levels

### PR DESCRIPTION
**Pull request by telezhnaya**
More about tx execution levels: https://docs.near.org/api/rpc/transactions#tx-status-result

The levels that allow having not all the execution outcomes (Included, IncludedFinal), actually had to wait for all of the execution outcomes anyway.
[1.37.3 had a separate partial fix for supporting Included status, but let's not overcomplicate]
With the current change, we can show everything we have at the moment without extra waiting.
Also, current change helps us to be 1 step closer to resolving https://github.com/near/nearcore/issues/6834
That would be my next step after we merge this PR.

TLDR: I partially copy pasted the code.
I could have rewritten `chain:get_final_transaction_result` instead, but it has LOTS of usages everywhere. I decided it's just not safe.
I anyway changed it a little with the new function `chain:get_execution_status` (hopefully in a better way) - see the details below in the comment.
But the new logic with partial execution outcomes list is used only in view_client (aka RPC, as far as I see from the usages).

### CODE REVIEW

File: chain/chain/src/chain.rs [hunk 1]
```diff
@@ -3077,6 +3077,39 @@ impl Chain {
...
+                    awaiting_receipt_ids.extend(&outcome.outcome.receipt_ids);
+                }
+            }
+        }
+        return if (&awaiting_receipt_ids - &seen_receipt_ids).is_empty() {
+            result
+        } else {
+            FinalExecutionStatus::Started
+        };
+    }
```
Comment (telezhnaya) [inline chain/chain/src/chain.rs:3110]: I've decided to extract this logic from `get_final_transaction_result` to a separate function, and rewrite it, fixing a number of potential bugs in the meantime.

Some important context:
The receipt may produce tons of new receipts. However, the corresponding execution outcome will have the status `ExecutionStatusView::SuccessReceiptId` with only one id inside - it's the last receipt id of all the produced receipts. So if you want to see the status, you have to go through all the list, it's not enough to look at the status of the id inside `SuccessReceiptId` field.

-----

Results

- I stopped relying on the order of the receipts
- I started checking all the produced receipts statuses, not only the last one
- I fixed a bug. Before, we have
```
match &outcome_with_id.outcome.status {
                         ExecutionStatusView::Unknown if num_outcomes == 1 => {
                             Some(FinalExecutionStatus::NotStarted)
                         }
                         ...
}

```
In reality, if we have 1 outcome, it means the status is at least Started.

**Review by telezhnaya — COMMENTED** [review]
The test https://github.com/near/nearcore/actions/runs/8530446007/job/23368253044?pr=10792
is failing, I have no idea what's wrong with it. I ran 2 times, both time failing, looks like it's not a glitch.

Please discuss it with me.
What's going on there:
Almost all the test runs flawlessly, we have issue at the very end.
We create an account with broadcast_tx_commit function
We check the result of the execution, everything is fine, no error
Then we ask for the account status, with optimistic finality, and we see that account does not exist
Do you have any idea why we can see this issue? Have you ever seen this test failing with the same error?
Feel free to ping me in Zulip/tg/slack, I'd be happy to chat about this

File: chain/client/src/view_client.rs [hunk 2]
```diff
@@ -507,7 +504,7 @@ impl ViewClientActor {
             target_shard_id,
             true,
         ) {
-            match self.chain.get_final_transaction_result(&tx_hash) {
+            match self.chain.get_transaction_result_for_rpc(&tx_hash) {
```
Comment (Longarithm) [inline chain/client/src/view_client.rs:507]: Consider renaming `get_tx_status` to `get_tx_status_for_rpc` to emphasize what is the purpose of the whole flow.
Or give names `get_partial_tx_status` and `get_partial_transaction_result` to make all methods consistent.

File: chain/chain/src/chain.rs [hunk 3]
```diff
@@ -3091,39 +3124,31 @@ impl Chain {
         Ok(())
     }
 
+    /// Collect all the execution outcomes existing at the current moment
+    fn get_partial_recursive_transaction_results(
```
Comment (Longarithm) [inline chain/chain/src/chain.rs:3128]: To avoid duplication, I would give `get_recursive_transaction_results` parameter like `fail_on_missing_outcome` (could it be shorter?), rename original function to `get_all_recursive_transaction_results` and just call `get_recursive_transaction_results(..., true)` there. `get_partial_recursive_transaction_results` would call `get_recursive_transaction_results(..., false)`.

File: chain/chain/src/chain.rs [hunk 4]
```diff
@@ -3077,6 +3077,39 @@ impl Chain {
...
+                        // let's continue sticking to it
+                        result = FinalExecutionStatus::SuccessValue(v.clone());
+                    }
+                }
+                ExecutionStatusView::SuccessReceiptId(_) => {
+                    awaiting_receipt_ids.extend(&outcome.outcome.receipt_ids);
+                }
+            }
+        }
+        return if (&awaiting_receipt_ids - &seen_receipt_ids).is_empty() {
```
Comment (Longarithm) [inline chain/chain/src/chain.rs:3105]: is_subset looks more readable. Looks like sets must be equal, but it doesn't matter.

File: integration-tests/src/user/runtime_user.rs [hunk 5]
```diff
@@ -194,6 +194,7 @@ impl RuntimeUser {
         transactions
     }
 
+    // todo get rid of tons of copy pasted code, it's usually outdated comparing to the original
```
Comment (Longarithm) [inline integration-tests/src/user/runtime_user.rs:197]: Add number of specific issue like `TODO(#1234)` so someone else could pick it up.

**Review by Longarithm — APPROVED** [review]
Thank you! Please also look at the comments.

Ideally we should have a test for new behaviour. We likely have tests which check that getting recursive outcomes fails at some early block and succeeds later. We can add checks there that getting partial outcomes always succeeds and gives ids which indeed correspond to spawned receipts.
